src/backend/distributed/planner/extended_op_node_utils.c

```diff
@@ -83,41 +81,86 @@ BuildExtendedOpNodeProperties(MultiExtendedOp *extendedOpNode)
 
 
 /*
- * GroupedByDisjointPartitionColumn returns true if the query is grouped by the
- * partition column of a table whose shards have disjoint sets of partition values.
+ * GroupedByPartitionColumn returns true if a GROUP BY in the opNode contains
+ * the partition column of the underlying relation, which is determined by
+ * searching the MultiNode tree for a MultiTable and MultiPartition with
+ * a matching column.
+ *
+ * When there is a re-partition join, the search terminates at the
+ * MultiPartition node. Hence we can push down the GROUP BY if the join
+ * column is in the GROUP BY.
  */
 static bool
-GroupedByDisjointPartitionColumn(List *tableNodeList, MultiExtendedOp *opNode)
+GroupedByPartitionColumn(MultiNode *node, MultiExtendedOp *opNode)
 {
-	bool result = false;
-	ListCell *tableNodeCell = NULL;
+	if (node == NULL)
+	{
+		return false;
+	}
 
-	foreach(tableNodeCell, tableNodeList)
+	if (CitusIsA(node, MultiTable))
 	{
-		MultiTable *tableNode = (MultiTable *) lfirst(tableNodeCell);
+		MultiTable *tableNode = (MultiTable *) node;
+
 		Oid relationId = tableNode->relationId;
 
-		if (relationId == SUBQUERY_RELATION_ID || !IsDistributedTable(relationId))
+		if (relationId == SUBQUERY_RELATION_ID ||
+			relationId == SUBQUERY_PUSHDOWN_RELATION_ID)
 		{
-			continue;
+			/* ignore subqueries for now */
+			return false;
 		}
 
 		char partitionMethod = PartitionMethod(relationId);
 		if (partitionMethod != DISTRIBUTE_BY_RANGE &&
 			partitionMethod != DISTRIBUTE_BY_HASH)
 		{
-			continue;
+			/* only range- and hash-distributed tables are strictly partitioned  */
+			return false;
 		}
 
 		if (GroupedByColumn(opNode->groupClauseList, opNode->targetList,
 							tableNode->partitionColumn))
 		{
-			result = true;
-			break;
+			/* only range- and hash-distributed tables are strictly partitioned  */
```

Comment:
does this comment belong to here?

Reply:
I don't think so. I lean towards removing these two comments without replacement